### PR TITLE
Copter: Guided Separate logging of Position_Targets and Attitude_Targets

### PR DESCRIPTION
**Pull request by shiv-tyagi**
This adds a new logger message to log information about attitude targets in guided mode. Earlier we were using the same GUID message to log information about both position and attitude targets in guided mode which is not good because GUID message did not have fields for yaw, roll and pitch (we were writing these values in pX, pY, pZ fields) and many other fields in the existing message are not needed for attitude target information.
Hence, we introduced a new message named GATT in copter for this purpose. We renamed the earlier message to GPOS which makes it more clear that this message is about POSITION target. The renaming is done for all vehicles for consistency.
I have tested this on SITL by sending following commands.

- `param set GUID_OPTIONS 8`
- `GUIDED`
- `arm throttle`
- `takeoff 10`

The following commands are to check if we get correct log messages for attitude targets.
- `attitude 0.9961947 0.0871557 0 0 0.5` (roll at 10 deg at 50% throttle)
- `attitude 0.9848078 0 0.1736482 0 0.4` (pitch at 20 deg at 40% throttle)
- `attitude 0.9659258 0 0 0.258819 0.3` (yaw at 30 deg at 30% throttle)
- `attitude 0.9515485 0.0381346 0.1893079 0.2392983 0.6` (roll 10 deg, pitch 20 deg and yaw 30 deg at 60% throttle`

- `module load message`

The following commands are to check if the existing message for position targets is working correctly (after renaming)
- `message SET_POSITION_TARGET_GLOBAL_INT 0 0 0 6 3576 -353621474 1491651746 10 0 0 0 0 0 0 0 0` (fly to lat,lon of -35.36,149.16 and 10m above home)
- `message SET_POSITION_TARGET_GLOBAL_INT 0 0 0 6 3527 0 0 0 1 0 0 0 0 0 0 0` (fly North at 1m/s)
- `message SET_POSITION_TARGET_GLOBAL_INT 0 0 0 6 2503 0 0 0 0 0 0 0 0 0 0.7854` (turn to North-East,  Yaw target + velocity of zero)

I checked the log and found all the messages correct.
Here are some screenshots.
Screenshot 2021-12-30 at 9 36 29 AM
Screenshot 2021-12-30 at 9 40 24 AM

Thanks @hendjoshsr71 for all your guidance and help in testing.

### CODE REVIEW

File: ArduCopter/Log.cpp
Comment (hendjoshsr71) [inline ArduCopter/Log.cpp:601]: Roll is defined -180deg to 180deg and pitch -90 to 90deg (though not exactly 90deg or bad things happen ;) )
The yaw is sent in as `guided_angle_state.yaw_cd = wrap_180_cd(ToDeg(guided_angle_state.yaw_cd) * 100.0f);` so we should use degrees `d` for that too

```suggestion
      "GATT",  "QBfffffff",    "TimeUS,Type,Roll,Pitch,Yaw,RollRate,PitchRate,YawRate,Thrust", "s-dddkkk-", "F-BBBBBB-" , true },
```

File: ArduCopter/Log.cpp
Comment (hendjoshsr71) [inline ArduCopter/Log.cpp:598]: We need to add climb_rate here too since the same field can be either thrust (-1,1) or interpreted as climb_rate. Then add it into the logger field and as its own input to the Log_Write_Guided_Attitude() too.
```suggestion
// @Field: Thrust: Thrust -1 to 1 
// @Field: Climb Rate
```

File: ArduCopter/mode_guided.cpp
Comment (hendjoshsr71) [inline ArduCopter/mode_guided.cpp:621]: Like I said above we should add this in. I prefer for it to be separate from thrust (though it wastes some bytes) as it makes log analysis easier by having units attached to it.
```suggestion
    copter.Log_Write_Guided_Attitude_Target(guided_mode, guided_angle_state.roll_cd, guided_angle_state.pitch_cd, guided_angle_state.yaw_cd, 0, 0, 0, guided_angle_state.thrust, guided_angle_state.climb_rate_cms);
```

**Review by hendjoshsr71 — CHANGES_REQUESTED** [review]
@shiv-tyagi  Thanks for this! I think this will be helpful when reviewing logs from folks to separate the control mode types. And not need to cleverly interpret the logs for each.

It will conflict a bit with @lthall 's PR  to fully use `set_attidue_target` but not in a horrible way.

File: ArduCopter/mode_guided.cpp
Comment (shiv-tyagi) [reply ArduCopter/mode_guided.cpp:621]: Done!!

File: ArduCopter/Log.cpp
Comment (shiv-tyagi) [reply ArduCopter/Log.cpp:598]: Done!

File: ArduCopter/Log.cpp
Comment (hendjoshsr71) [inline ArduCopter/Log.cpp:599]: The unit type should be handled by the logger fields themselves. To make an output in degrees/sec. These comments are actually wrong since with the multiplier they are in deg/sec.

```suggestion
// @Field: RollRt: Roll rate
// @Field: PitchRt: Pitch rate
// @Field: YawRt: TYaw rate
```

**Review by hendjoshsr71 — CHANGES_REQUESTED** [review]

File: ArduCopter/Log.cpp
Comment (lthall) [reply ArduCopter/Log.cpp]: I would say, convert everything to degrees/s before logging and leave the "in degrees/s"

Comment (shiv-tyagi) [reply ArduCopter/Log.cpp]: @lthall I left them in centidegrees because in the former GUID message too we were logging values in centimeters, centidegrees etc. So I tried to maintain that consistency. If you still say, I would be happy to convert these :-).

Comment (lthall) [reply ArduCopter/Log.cpp]: I would strongly support moving to degrees. If we are going to change them lets do it properly.

Comment (shiv-tyagi) [reply ArduCopter/Log.cpp]: Sure! I will change all units to degrees and deg/s wherever required. Thanks for your suggestion.

File: ArduCopter/mode_guided.cpp
Comment (hendjoshsr71) [inline ArduCopter/mode_guided.cpp:621]: Always prefer multiplication to division (where possible) division takes longer on the hardware.  We might as well do the same for climb_rate as well.

```suggestion
    copter.Log_Write_Guided_Attitude_Target(guided_mode, guided_angle_state.roll_cd * 0.01, guided_angle_state.pitch_cd * 0.01, guided_angle_state.yaw_cd * 0.01, 0, 0, 0, guided_angle_state.thrust, guided_angle_state.climb_rate_cms * 0.01);
```

File: ArduCopter/Log.cpp
Comment (hendjoshsr71) [inline ArduCopter/Log.cpp:637]:
```suggestion
void Copter::Log_Write_Guided_Attitude_Target(ModeGuided::SubMode target_type, float roll, float pitch, float yaw, float roll_rate, float pitch_rate, float yaw_rate, float thrust, float climb_rate) {}
```

File: ArduCopter/Log.cpp
Comment (hendjoshsr71) [inline ArduCopter/Log.cpp:422]:
```suggestion
// roll, pitch and yaw are in degrees
// roll_rate, pitch_rate and yaw_rate are in degrees per second
// thrust is between -1 to 1
// climb_rate is in (m/s)
```

File: ArduCopter/Log.cpp
Comment (hendjoshsr71) [inline ArduCopter/Log.cpp:436]:
```suggestion
void Copter::Log_Write_Guided_Attitude_Target(ModeGuided::SubMode target_type, float roll, float pitch, float yaw, float roll_rate, float pitch_rate, float yaw_rate, float thrust, float climb_rate)
{
    struct log_Guided_Attitude_Target pkt = {
        LOG_PACKET_HEADER_INIT(LOG_GUIDED_ATTITUDE_TARGET_MSG),
        time_us         : AP_HAL::micros64(),
        type            : (uint8_t)target_type,
        roll            : roll,
        pitch           : pitch,
        yaw             : yaw,
        roll_rate       : roll_rate,
        pitch_rate      : pitch_rate,
        yaw_rate        : yaw_rate,
        thrust          : thrust,
        climb_rate      : climb_rate
```

**Review by hendjoshsr71 — CHANGES_REQUESTED** [review]
Need to correct thing to all be in degrees now. As comments and variables names say they are in `zzz_cds`

File: ArduCopter/mode_guided.cpp
Comment (shiv-tyagi) [reply ArduCopter/mode_guided.cpp:621]: Done! Thanks

File: ArduCopter/Log.cpp
Comment (shiv-tyagi) [reply ArduCopter/Log.cpp:637]: Oops sorry! I forgot to change these

File: ArduCopter/Log.cpp
Comment (hendjoshsr71) [inline ArduCopter/Log.cpp:605]: Nitpick:
The difference between `-` and `0` is subtle. The first means a multiplier does not apply and the other means `x1`. The result is the same code wise, but interpretation is different. 

```suggestion
    { LOG_GUIDED_ATTITUDE_TARGET_MSG, sizeof(log_Guided_Attitude_Target),
      "GATT",  "QBffffffff",    "TimeUS,Type,Roll,Pitch,Yaw,RollRt,PitchRt,YawRt,Thrust,ClimbRt", "s-dddkkk-n", "F-000000-B" , true },
```

**Review by hendjoshsr71 — APPROVED** [review]
LGTM

**Comment by tridge** [comment]
nice idea!
needs review by @rmackay9

**Comment by shiv-tyagi** [comment]
It was @hendjoshsr71 's idea. Thanks to him :-)

**Comment by hendjoshsr71** [comment]
@shiv-tyagi  Now that we have the set_attitude PR, https://github.com/ArduPilot/ardupilot/pull/18833,  is in this one needs a rebase. Let me know if you need help doing that?

**Comment by shiv-tyagi** [comment]
@hendjoshsr71 I have rebased. Please check.

File: ArduCopter/mode_guided.cpp
Comment (hendjoshsr71) [inline ArduCopter/mode_guided.cpp:624]: You have lost the angular velocity here (ang_vel). angular_velocity = [roll_rate, pitch_rate, yaw_rate]

Make this pass by const reference instead in the the Logging function.

```suggestion
    copter.Log_Write_Guided_Attitude_Target(guided_mode, degrees(roll_rad), degrees(pitch_rad), degrees(yaw_rad), ang_vel, guided_angle_state.thrust, guided_angle_state.climb_rate_cms * 0.01);
```

File: ArduCopter/Copter.h
Comment (hendjoshsr71) [inline ArduCopter/Copter.h:799]:
```suggestion
    void Log_Write_Guided_Attitude_Target(ModeGuided::SubMode target_type, float roll, float pitch, float yaw, const Vector3f &ang_vel, float thrust, float climb_rate);
```

File: ArduCopter/Log.cpp
Comment (hendjoshsr71) [inline ArduCopter/Log.cpp:399]:
```suggestion
void Copter::Log_Write_Guided_Attitude_Target(ModeGuided::SubMode target_type, float roll, float pitch, float yaw, const Vector3f &ang_vel, float thrust, float climb_rate)
```

File: ArduCopter/Log.cpp
Comment (hendjoshsr71) [inline ArduCopter/Log.cpp:410]:
```suggestion
        roll_rate       : ang_vel.x,
        pitch_rate      : ang_vel.y,
        yaw_rate        : ang_vel.z,
```

File: ArduCopter/Log.cpp
Comment (hendjoshsr71) [inline ArduCopter/Log.cpp:601]:
```suggestion
void Copter::Log_Write_Guided_Attitude_Target(ModeGuided::SubMode target_type, float roll, float pitch, float yaw, const Vector3f &ang_vel, float thrust, float climb_rate) {}
```

**Review by hendjoshsr71 — CHANGES_REQUESTED** [review]
You have lost the angular velocity (ang_vel). 
angular_velocity = [roll_rate, pitch_rate, yaw_rate]

Make this pass by const reference instead in the the Logging function.

**Review by hendjoshsr71 — CHANGES_REQUESTED** [review]
one last thing i think.  

Give it a test too if you haven't done so since you coded this initially.

**Comment by shiv-tyagi** [comment]
```2022-01-23 18:03:17.791 GATT {TimeUS : 72975798, Type : 6, Roll : 9.999994277954102, Pitch : 0.0, Yaw : 0.0, RollRt : 0.0, PitchRt : 0.0, YawRt : 0.0, Thrust : 0.5, ClimbRt : 0.0}
2022-01-23 18:03:44.014 GATT {TimeUS : 99198638, Type : 6, Roll : 0.0, Pitch : 20.000001907348633, Yaw : 0.0, RollRt : 0.0, PitchRt : 0.0, YawRt : 0.0, Thrust : 0.4000000059604645, ClimbRt : 0.0}
2022-01-23 18:04:01.964 GATT {TimeUS : 117148955, Type : 6, Roll : 0.0, Pitch : 0.0, Yaw : 29.999998092651367, RollRt : 0.0, PitchRt : 0.0, YawRt : 0.0, Thrust : 0.30000001192092896, ClimbRt : 0.0}
2022-01-23 18:04:36.259 GATT {TimeUS : 151443565, Type : 6, Roll : 10.000003814697266, Pitch : 20.000003814697266, Yaw : 29.999998092651367, RollRt : 0.0, PitchRt : 0.0, YawRt : 0.0, Thrust : 0.6000000238418579, ClimbRt : 0.0}
```
```
2022-01-23 18:05:06.121 GPOS {TimeUS : 181305782, Type : 2, pX : -353621472.0, pY : 1491651712.0, pZ : 1000.0, Terrain : 0, vX : 0.0, vY : 0.0, vZ : 0.0, aX : 0.0, aY : 0.0, aZ : 0.0}
2022-01-23 18:05:49.416 GPOS {TimeUS : 224600957, Type : 4, pX : 0.0, pY : 0.0, pZ : 0.0, Terrain : 0, vX : 100.0, vY : 0.0, vZ : -0.0, aX : 0.0, aY : 0.0, aZ : 0.0}
2022-01-23 18:07:41.559 GPOS {TimeUS : 336743582, Type : 4, pX : 0.0, pY : 0.0, pZ : 0.0, Terrain : 0, vX : 0.0, vY : 0.0, vZ : -0.0, aX : 0.0, aY : 0.0, aZ : 0.0}
```
@hendjoshsr71 
I have retested this (with same commands mentioned at the beginning of this thread) and all the messages seem to be good.

File: ArduCopter/Log.cpp
Comment (shiv-tyagi) [reply ArduCopter/Log.cpp]: fixed it

**Review by hendjoshsr71 — APPROVED** [review]
LGTM

File: ArduCopter/Log.cpp
Comment (peterbarker) [inline ArduCopter/Log.cpp:376]:
```suggestion
    const struct log_Guided_Position_Target pkt {
```

File: ArduCopter/Log.cpp
Comment (peterbarker) [inline ArduCopter/Log.cpp:339]:
```suggestion
const struct PACKED log_Guided_Position_Target {
```

File: ArduCopter/Log.cpp
Comment (peterbarker) [inline ArduCopter/Log.cpp:401]:
```suggestion
    const struct log_Guided_Attitude_Target pkt {
```

**Review by peterbarker — CHANGES_REQUESTED** [review]

**Review by rmackay9 — APPROVED** [review]
Hi @shiv-tyagi, thanks very much for this.  Could you rename GPOS to GUIP and GATT to GUIA?

**Review by rmackay9 — CHANGES_REQUESTED** [review]
Oops, sorry, I marked as approved when I meant to mark, "Request changes".  In general this looks good though so I intend to merge after the name change.  txs again!

File: ArduCopter/Log.cpp
Comment (shiv-tyagi) [reply ArduCopter/Log.cpp]: Done

File: ArduCopter/Log.cpp
Comment (shiv-tyagi) [reply ArduCopter/Log.cpp]: Done

File: ArduCopter/Log.cpp
Comment (shiv-tyagi) [reply ArduCopter/Log.cpp]: Done

**Review by hendjoshsr71 — CHANGES_REQUESTED** [review]
Looks good besides the nitpicks

File: ArduCopter/Log.cpp
Comment (shiv-tyagi) [reply ArduCopter/Log.cpp:396]: Fixed it.

File: ArduCopter/Log.cpp
Comment (shiv-tyagi) [reply ArduCopter/Log.cpp:371]: fixed it. Gave it a test too. Working good.

**Review by hendjoshsr71 — APPROVED** [review]

**Comment by rmackay9** [comment]
Merged, thanks @shiv-tyagi!

**Comment by hendjoshsr71** [comment]
Thanks @shiv-tyagi ! This will make parsing logs a lot easier going forward!

**Pull request merged**
